docs/rfcs/rfc-006-auth-plugin.md

```diff
@@ -0,0 +1,98 @@
+# Auth Plugin
+
+## Problem
+
+Neo4j GraphQL uses the [Node.js `crypto`](https://nodejs.org/api/crypto.html) library to decode an incoming JWT, and If this usage could be extracted into a 'plugin' it would path the way for future developer tools surrounding Neo4j GraphQL such as Neo4j Desktop interoperability.
+
+## Proposed Solution
+
+Extract the current auth decoding functionally into a new package `@neo4j/graphql-plugin-auth`, this package would be deployed independently to NPM.
```

Comment:
You may call me ridiculous for this, but I actually think we need 3 plugins here, or at least 3 classes.

Each using a slight twist on the abstract class proposed below, with the following changes:

- Added constructor with default behaviour
- Removing `secret`
- Adding `rolesPath`
- Making `decode` abstract
- Changing `Record<string, unknown>>` to `JwtPayload`, and also adding `undefined` return

```ts
abstract class Neo4jGraphQLAuth {
    rolesPath: string; 

    constructor(rolesPath: string) {
        this.rolesPath = rolesPath;
    }

    abstract decode<T = JwtPayload | undefined>(jwt: string): Promise<T>;
}
```

## `Neo4jGraphQLAuthNoVerify`

Decodes an already verified JWT. Uses the base constructor from the abstract class.

```ts
abstract class Neo4jGraphQLAuthNoVerify extends Neo4jGraphQLAuth {
    decode<T = JwtPayload | undefined>(jwt: string): Promise<T> {
        return jsonwebtoken.decode(jwt, { json: true }) || undefined;
    }
}
```

## `Neo4jGraphQLAuthSecret`

Verifies the JWT with the provided secret. Overloads the default constructor, accepting a `secret` argument.

```ts
abstract class Neo4jGraphQLAuthSecret extends Neo4jGraphQLAuth {
    secret: string;

    constructor(secret: string, rolesPath: string) {
        super(rolesPath);
        this.secret = secret;
    }

    decode<T = JwtPayload | undefined>(jwt: string): Promise<T> {
        return jsonwebtoken.verify(jwt, this.secret, {
            algorithms: ["HS256", "RS256"],
        });
    }
}
```

## `Neo4jGraphQLAuthJWKS`

Verifies the JWT with a JWKS endpoint. Overloads the default constructor, accepting a `endpoint` argument.

```ts
abstract class Neo4jGraphQLAuthJWKS extends Neo4jGraphQLAuth {
    endpoint: string;

    constructor(endpoint: string, rolesPath: string) {
        super(rolesPath);
        this.endpoint = endpoint;
    }

    decode<T = JwtPayload | undefined>(jwt: string): Promise<T> {
        // Do the JWKS endpoint work
    }
}
```

Reply:
It may seem like a lot, but the separation of concerns will make each plugin serve one purpose and one purpose only and much easier to document.

Reply:
I don't think splitting the current auth functionally into three separate classes is necessary! This would mean debugging would be difficult due to the fact, instead of just setting `noVerify: true`, I would now have to instantiate a new class whether I'm verifying or temporary turning the verification off.  I believe that using a single class and exposing all the existing auth options on it is fine! 

**Existing Options**

```ts
export interface Neo4jGraphQLJWT {
    jwksEndpoint?: string;
    secret?: string;
    noVerify?: boolean;
    rolesPath?: string;
}
```

**Class Options**

```ts
export interface Neo4jGraphQLAuthOptions {
    jwksEndpoint?: string;
    secret?: string;
    noVerify?: boolean;
    // rolesPath?: string; ## This should be applied to Neo4jGraphQL as the auth plugin dont need it!
}

abstract class Neo4jGraphQLAuth {
    options: Neo4jGraphQLAuthOptions;

    constructor(o: Neo4jGraphQLAuthOptions) {
        this.options = o;
    }
}
```

Note that the auth decoding doesn't need `rolesPath` so this should be supplied in the `Neo4jGraphQL` constructor.

Reply:
Upon reflection, you're right about `noVerify`, however, I strongly stand behind JWKS functionality being a separate plugin to just decoding using a secret. Especially given that `noVerify` is not an option when using JWKS, so having them all together results in redundant arguments for certain situations.

> Note that the auth decoding doesn't need `rolesPath` so this should be supplied in the `Neo4jGraphQL` constructor.

I definitely think this should be in the plugin, otherwise auth functionality is split across multiple instantiations rather than just one.